libs/utils/lingo.js

```diff
@@ -0,0 +1,70 @@
+async function urlInMatchingIndex(matchingIndexes, sanitizedPath) {
+  const pathsArrays = await Promise.all(matchingIndexes.map((q) => q.pathsRequest));
+  const allPaths = pathsArrays.flat().filter(Boolean);
+  return allPaths.some((path) => sanitizedPath === path);
+}
+
+async function tryEarlyDecisionUsingBaseIndex(
+  matchingIndexes,
+  regionalPath,
+  basePath,
+  urlHostname,
+  initialResolvedCount,
+  baseQueryIndex,
+) {
+  if (!baseQueryIndex?.pathsRequest) return null;
+
+  const baseIndexMatches = baseQueryIndex.domains.includes(urlHostname) ? [baseQueryIndex] : [];
+  if (baseIndexMatches.length === 0) return null;
+
+  const allRegionalPromises = Promise.all(matchingIndexes.map((m) => m.pathsRequest));
+
+  await Promise.race([baseQueryIndex.pathsRequest, allRegionalPromises]);
+
+  if (!baseQueryIndex.requestResolved) return null;
+
+  const currentResolvedIndexes = matchingIndexes.filter((m) => m.requestResolved);
+
+  if (currentResolvedIndexes.length > initialResolvedCount) {
+    const foundInNewlyResolved = await urlInMatchingIndex(currentResolvedIndexes, regionalPath);
+    if (foundInNewlyResolved) return true;
+  }
+
+  // Check if basePath exists in base query index
+  const urlExistsInBase = await urlInMatchingIndex(baseIndexMatches, basePath);
+  return urlExistsInBase ? false : null;
+}
```

Comment:
IMO you can shorten this a bit, i highlighted my thinking inline, feel free to kill the review comments in `//` 
```suggestion
async function tryEarlyDecisionUsingBaseIndex(
  matchingIndexes,
  regionalPath,
  basePath,
  urlHostname,
  initialResolvedCount,
  baseQueryIndex = {},
) {
  // Combine early return, save the variable makes it a bit leaner
  if (!baseQueryIndex.pathsRequest || !baseQueryIndex.domains.includes(urlHostname)) return null;

  // Make this a bit leaner
  await Promise.race([
    baseQueryIndex.pathsRequest,
    Promise.all(matchingIndexes.map((m) => m.pathsRequest))
  ]);

  if (!baseQueryIndex.requestResolved) return null;

  const currentResolvedIndexes = matchingIndexes.filter((m) => m.requestResolved);
  
  if (currentResolvedIndexes.length > initialResolvedCount) {
    const foundInNewlyResolved = await urlInMatchingIndex(currentResolvedIndexes, regionalPath);
    if (foundInNewlyResolved) return true;
  }

  // [baseQueryIndex] will be sure to exist, since we had the early return otherwise
  const urlExistsInBase = await urlInMatchingIndex([baseQueryIndex], basePath);
  return urlExistsInBase ? false : null;
}
```